src/main/java/com/pnudev/librarysystem/service/BorrowingService.java

```diff
@@ -0,0 +1,77 @@
+package com.pnudev.librarysystem.service;
+
+import com.pnudev.librarysystem.dto.BorrowingDTO;
+import com.pnudev.librarysystem.dto.CreateBorrowingDTO;
+import com.pnudev.librarysystem.entity.Book;
+import com.pnudev.librarysystem.entity.Borrowing;
+import com.pnudev.librarysystem.entity.User;
+import com.pnudev.librarysystem.enums.BorrowingStatus;
+import com.pnudev.librarysystem.exception.OperationFailedException;
+import com.pnudev.librarysystem.mapper.BorrowingMapper;
+import com.pnudev.librarysystem.repository.BorrowingRepository;
+import com.pnudev.librarysystem.security.UserDetailsImpl;
+import jakarta.persistence.EntityNotFoundException;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDate;
+import java.util.List;
+
+@RequiredArgsConstructor
+@Service
+public class BorrowingService {
+    private final UserService userService;
+    private final BookService bookService;
+    private final BorrowingRepository borrowingRepository;
+    private final BorrowingMapper borrowingMapper;
+
+    @Value("${application.reservation.expiration-in-days}")
+    private int reservationExpirationInDays;
+
+
+    public void reserveBook(UserDetailsImpl userDetailsImpl, CreateBorrowingDTO createBorrowingDTO) {
+        Long userId = userDetailsImpl.getId();
+        Long bookId = createBorrowingDTO.getBookId();
+
+        if (!bookService.isBookAvailable(bookId)) {
+            throw new OperationFailedException("Book is not available");
+        }
+
+        if (!userService.canUserReserve(userId)) {
+            throw new OperationFailedException("User has reached the maximum number of reservations");
+        }
+
+        Borrowing borrowing = Borrowing.builder()
+                .user(new User(userId))
```

Comment:
It would be better if your `User` already implemented `UserDetails`, that way you would be able to just grab it from `SecurityContextHolder` as-is. Because you are basically creating a new user for each and every borrowing.
Alternatively you can just extend `UserService` with a method to get user from `userId` and passing it in this builder

Reply:
Changed to an additional id field option